tests/compiler/rt/finalize.ts

```diff
@@ -0,0 +1,55 @@
+class FinalizationRegistry {
+  private map: Map<usize,string> = new Map();
+
+  constructor(
+    private fn: (value: string) => void
+  ) {}
+
+  register<U>(target: U, value: string): void {
+    if (!isReference<U>()) throw new Error("not an object");
+    this.map.set(changetype<usize>(target), value);
+  }
+
+  @unsafe finalize(ptr: usize): void {
+    var map = this.map;
+    if (map.has(ptr)) {
+      let value = map.get(ptr);
+      this.fn(value);
+      map.delete(ptr);
+    }
+  }
+}
+
+var registry = new FinalizationRegistry(value => {
```

Comment:
So this is how we should hook into object finalization? Any tips or important things we should be aware of? Also, where should this go in the documentation?

Reply:
That's really just an initial idea along the lines of [JS weakrefs](https://github.com/tc39/proposal-weakrefs). The advantage this has over destructors is that the finalization function isn't called with the actual object being free'd, which would be super unsafe, but with the registered value. Nothing set in stone yet.